### ISSUE TITLE
Add tag to declare variable types

### ISSUE THREAD

**Issue by drjayvee** [twigphp/Twig#4165]
See #4003 and #4009 for some context. In short, since the PHP language itself as well as its community are moving to (optional) stricter typing, it makes sense for Twig to support this as well by adding a `types` tag.

~~@fabpot said he'd like to bootstrap a spec. This is the first draft. Everything is up for debate, obviously. In addition, I've put in italics parts of the spec that I have doubts about myself.~~

Given the discussions below (thanks all for contributing), I think the spec is now crystallizing. Consider it a beta.

# Syntax
The `types` tag contains a mapping of variable name to constant type string (no expression or interpolation):

```
{% types {variable: 'Type'} %}
```
The reason for quoting the type is that several characters (e.g., `\`) are currently not supported by the Lexer. It also allows analyzer implementations to freely specify the syntax.

The variable name _must not_ be quoted. This is to ensure that only valid names can be used.

# Types
The syntax and content of the types are not part of this specification. This provides complete freedom to extensions that inspect and analyze the tags. It also keeps the implementation in Twig very simple and robust.

(In practice, it's likely that implementations will rely on widely-used specifications like [PHPDoc / PHPStan types](https://phpstan.org/writing-php-code/phpdoc-types).)

## Examples
Here are two examples:
```twig
{% types {foo: 'bool'} %}

{% types {
  bar: "list<int>",
  baz: '?\Name\Space\ClassName'
} %}
```

# Behavior
This specification does require any behavior other than validating the syntax for the tag itself. The tag does not lead to generated PHP code.

# Usage
This section describes possible implementations to demonstrate the utility of this addition.

It is up to extensions developed by the community to implement any of this.

## Development time
The types can be useful during development by enabling IDEs to auto-complete code, issue warnings, and more.

This is [already implemented](https://github.com/twigphp/Twig/pull/4009#issuecomment-2003568293) by the Symfony Support plugin for PHPStorm when using comments to declare variable types (e.g., `{# @var id int #}`).

These tags also provide precious documentation about the variables expected in a template's context.

## Compile time
During compilation, a Twig extension can:
1. Check the type and throw an appropriate Exception if it is not valid.
2. For objects, check whether properties and more methods exist, and throw an appropriate Exception if there is none.
2. Add run-time checks (see below).

## Run time
During run time, behavior may depend on `strict_variables` environment option:
* If *disabled* (likely in production environments), a PHP warning is triggered.
* If *enabled*, (development and testing environments), an Exception is thrown.

## Possible extensions
* Use property access or method call for objects instead of `twig_get_attribute()` to improve performance. (see #3928)

**Comment by JoshuaBehrens** [comment]
Would you mind linking
> Use property access or method call for objects instead of twig_get_attribute() to improve performance

with https://github.com/twigphp/Twig/pull/3928 so we can see, when this is unblocked by this issue? :)

**Comment by ruudk** [comment]
If I may suggest: instead of deciding the possible var type values, why not accept any valid PHPDoc type. Together with [PHPStan's PHPDoc Parser](https://github.com/phpstan/phpdoc-parser) we can parse them into PhpDocNode's that then can be transformed to PHPStan types. From there, it can be used to build compile time assertions. Doing it from scratch feels like reinventing the wheel 😅 

We could still only accept a subset from the possible types, and let others improve it over time via PR's.

**Comment by ericmorand** [comment]
Give it enough time and Twig will end up being PHTML, the format it was supposed to replace.

**Comment by drjayvee** [comment]
> If I may suggest: instead of deciding the possible var type values, why not accept any valid PHPDoc type

Thanks, that's a good suggestion. I was mindful of not reinventing the wheel and hoped that we could re-use parts of PHPStan to parse and evaluate the types. Should we perhaps go all the way and support any PHPStan type? I'll add a sentence in italics to that section for now.

I'll add support for nullable types, which I forgot. But the more gets added, the more sense it makes to just go all the way

**Comment by willrowe** [comment]
I'm not sure if this would be the best place to implement this, but a lot of times I have a flag in a template to allow certain markup to be turned on or off, but need to retain previous functionality for other templates that are already including it. I usually have to add a `set` tag to check if the variable exists and set it to a default value if not. I already include a comment at the top of the template denoting what variables are used in the template, but it would be nice to be able to provide a default value for any optional variables in the same place that I denote the type, replacing the need for a bunch of `set` tags.

**Comment by ruudk** [comment]
```twig
{% var variable Type %}
```

I think that the order should be the [same as with inline PHPDocs](https://phpstan.org/writing-php-code/phpdocs-basics#inline-%40var):

```twig
{% var Type variable %}
```

**Comment by willrowe** [comment]
I agree with @ruudk on the order. It would also match how class properties and argument types are defined in PHP.

Also, going along with that and clarifying what I was proposing in https://github.com/twigphp/Twig/issues/4165#issuecomment-2265852945, this is what it would look like to specify a default value:
```twig
{% var Type variable = 'default_value' %}
{% var bool flag = false %}
```

**Comment by drjayvee** [comment]
> I think that the order should be the [same as with inline PHPDocs](https://phpstan.org/writing-php-code/phpdocs-basics#inline-%40var):

Agreed. I've updated the OP. (While I prefer the original order, we should follow existing conventions.)

We could also consider an optional description. [phpDocumentor supports this](https://docs.phpdoc.org/2.9/references/phpdoc/tags/var.html#), but PHPStan doesn't seem to.
```
{% var bool foo to foo or not to foo %}
```

Then again, this makes the tag less readable, plus there's nothing Twig itself would do with this. The alternative is to just add a comment (before or afterwards) for humans:
```
{% var bool foo %} {# to foo or not to foo #}
```

**Comment by ruudk** [comment]
> We could also consider an optional description. [phpDocumentor supports this](https://docs.phpdoc.org/2.9/references/phpdoc/tags/var.html#), but PHPStan doesn't seem to.

What's the point of providing a description there? 
The reason why PHPStan doesn't support it at that position is that the `@var` is already inside a comment. So you can add a comment above the `@var` to explain the why.

Same can be done with Twig:
```twig
{# I need to explain this somehow #}
{% var 'bool' foo %}
```

PS: The Type should be a string (ConstantStringExpression).

The variable could be a NameExpression or ConstantStringExpression.

**Comment by drjayvee** [comment]
> it would be nice to be able to provide a default value for any optional variables in the same place that I denote the type, replacing the need for a bunch of set tags.

I personally actually disagree on principle: declaring the type and any value should not be mixed. This does bring up two interesting questions, though.

First, should we add `undefined` as a possible type? Or should template developers ensure variables are always set (possibly using `null`)? Since `var` is all about stricter typing, I would prefer the latter.

Second, what do we do about variables defined using `set`? Should users add `var` afterwards? Or should Twig try to auto-detect the type? That would be very difficult (or outright infeasible) since `set` can use an expression (using literals and even function calls).

**Comment by drjayvee** [comment]
> What's the point of providing a description there? The reason why PHPStan doesn't support it at that position is that the `@var` is already inside a comment. So you can add a comment above the `@var` to explain the why.

Exactly. I was just thinking out loud.

> PS: The Type should be a string (ConstantStringExpression).
> 
> The variable could be a NameExpression or ConstantStringExpression.

Why should the type be quoted? It's not an expression (although I confess not to know the definition of "expression"). It also looks really weird to me and is inconsistent with PHPStan.

Do you think the implementation could be tricky if the type isn't quoted?

Update: the implementation does indeed run into issues. Therefore, I've amended the spec to quote types.

**Comment by willrowe** [comment]
> I personally actually disagree on principle: declaring the type and any value should not be mixed.

@drjayvee can you explain more?

**Comment by drjayvee** [comment]
> > I personally actually disagree on principle: declaring the type and any value should not be mixed.
> 
> @drjayvee can you explain more?

@willrowe, absolutely! I should have done that straight away.

But first off, I want to point out that this is just my personal opinion. I'd love to hear what other people think.

One reason I dislike your proposal is that Twig already has three ways of setting values:
1. injecting them into the template's context (in PHP)
2. using the `set` tag
3. defining a macro argument

I feel like adding a fourth way leads to bloat. (Or worse, conflicts.)

Macros arguments already have support for default values. (In fact, they're never even undefined, since they get an implicit default of `null`. I'm actually proposing to make this distinction clear in the AST: #4010)

The `set` tag already supports your use case just fine:
```twig
{% set foo = foo ?? 'bar' %}

{# or, in case null must be kept as-is, more strictly: #}
{% set foo = foo is defined ? foo : 'bar' %}
```

The second problem is that it feels like mixing concerns. I'm envisioning the `var` tag to be analogous to PHPStan's `@var` annotation, not PHP's native type declarations (e.g., for class properties and function arguments). Another way of looking at `var` is that it only affects template code (and therefore provide value) _after_ the tag. You can read it like "from here on out, foo is guaranteed to be of this type". 

However, I'm not super convinced by my own argument here, since both PHP and Typescript (as just two examples) do let you declare types and default values in one go in many places (e.g., function default values and class properties).


A third problem is that `var` is currently envisioned to have no effect on compiled templates. I think this is a sensible design goal is prevent compatibility issues and make the change more palatable to maintainers and users. If `var` does support default values, it will have to affect compilation output.


I hope I explained clearly. I'm sincerely curious as to you what you think now you know my reasoning.

**Comment by ericmorand** [comment]
@drjayvee 

> {% var ?\Name\Space\ClassName baz %}

What does it mean? What is `\Name\Space\ClassName` and how is such a type supposed to be declared?

**Comment by willrowe** [comment]
@drjayvee those are all great points and illustrate how you see this feature being utilized.

Where we differ is that I think about this as more of a dockblock/argument list at the top of the template file (similar to the `extends` tag) defining what variables should be available in the context of the template file. As if the file is a function and the `var` tags are its arguments. That's why it makes sense to me to have support for default value, since arguments can have default values.

Keeping the above in mind:
> One reason I dislike your proposal is that Twig already has three ways of setting values:
> 1. injecting them into the template's context (in PHP)

I see this as augmenting the template context specifically, so if a `var` is missing from the template context it will be initialized with the default value, if one has been specified.

> 2. using the set tag

Since I am thinking of this as a single definition at the top, it would be different than `set`. The `set` tag can be used multiple times throughout the template to change a variable value at runtime.

> 3. defining a macro argument

I think macros are separate from what I am talking about.

> I feel like adding a fourth way leads to bloat.

I don't see it as an additional way to set variable values within the template if it were to only affect the context as it comes into the template. It would be overridden by any values provided by the user.

> Macros arguments already have support for default values. (In fact, they're never even undefined, since they get an implicit default of null.

I'm not talking about macros, just template file context.

> The set tag already supports your use case just fine

As I said in my original comment, I already do this, but am looking for a way to easily combine the documentation of variables within a template file and a default value.

> The second problem is that it feels like mixing concerns. I'm envisioning the var tag to be analogous to PHPStan's @var annotation, not PHP's native type declarations (e.g., for class properties and function arguments). Another way of looking at var is that it only affects template code (and therefore provide value) after the tag. You can read it like "from here on out, foo is guaranteed to be of this type".

This is where we are thinking about the tag differently. It seems like adding the ability specify a type to the existing `set` tag would accomplish what you are describing here.

> A third problem is that var is currently envisioned to have no effect on compiled templates. I think this is a sensible design goal is prevent compatibility issues and make the change more palatable to maintainers and users. If var does support default values, it will have to affect compilation output.

I think it would be much more powerful to have the variable type information available after compiling, since that would allow introspection of what variables a template file expects to be passed as context, among other things. Compiling the types could also possibly allow Twig to leverage built-in PHP type checks so that checks don't necessarily need to be run manually at runtime, but would rather be implicitly checked.

All this being said, I think it comes down to whether it would be used as a way to define variable types once for the whole file, or inline.

**Comment by drjayvee** [comment]
> @drjayvee
> 
> > {% var ?\Name\Space\ClassName baz %}
> 
> What does it mean? What is `\Name\Space\ClassName` and how is such a type supposed to be declared?

This tag declares that variable `baz` must be an object of class with FQN `\Name\Space\ClassName` or null (due to the `?` prefix).

The PHP equivalent for a function argument would be
```php
function bar(?\Name\Space\ClassName $baz) {}
```

**Comment by drjayvee** [comment]
@willrowe thanks for your detailed response. It does indeed appear we have slightly different use cases in mind.

Documenting / specifying the template's context variables similar to function arguments is an explicit goal of this proposal. 
I still maintain that setting default values would be supported just fine with a combination of `set` and `var` (to declare type).

The current proposal is more generic: you can use `var` in macros and anywhere inside a template.

Consider the following example:
```twig
{% if user.isLoggedIn() %}
  {% set userName = user.fullName %} {# Class property is ?string but guaranteed to be string if user is logged in #}
{% else %}
  {% set userName = 'Anonymous' %}
{% endif %}

{% var string userName %}
```

The `var` tag on the last line serves as documentation and safety net. It guarantees that, `userName` is always `string`, even though `user.fullName` itself is nullable.

That being said, I very much would like the community to weigh in here.

Speaking of: @fabpot, you mentioned you wanted to involve some interested parties, no? I'd also like to know your thoughts on the proposal so far.

**Comment by ericmorand** [comment]
@drjayvee thanks. How do you declare to the twig compiler that this class exist? Said differently, at compile-time, how is the compiler expected to check that this type exist?

**Comment by drjayvee** [comment]
> @drjayvee thanks. How do you declare to the twig compiler that this class exist? Said differently, at compile-time, how is the compiler expected to check that this type exist?

Simply by using `class_exists()`, right? I guess the only potential issue is that autoloading must be enabled. But who _doesn't_ have that.

Do you see any other problem? Or am I overlooking something here?

**Comment by ericmorand** [comment]
@drjayvee you are explaining how the php implementation would work. But how would the other implementations work? More generally, from a specification point of view, how would `\Foo\Bar` be resolved?

**Comment by ericmorand** [comment]
@drjayvee also it is fundamental to establish how type checking would work? Let's say the a variable `x` of type `\Foo\Bar` is used this way:

```
{{ x.y.z }}
```

What would be the algorithm that compilers are supposed to implement to check that `x.y.z` is a valid member access? 

As I said earlier in this thread in an ironic manner, I'm concerned that twig would become a proper typed programming language, which means that it would also come with a complexity in both specification and implementation that the TwigPHP team may not be ready to face. I recommend that you have a look at how typed languages are specified and how their compilers are implemented before making a decision about this feature: this is very very complex topic.

**Comment by willrowe** [comment]
> I still maintain that setting default values would be supported just fine with a combination of set and var (to declare type).

- As I said, I already use `set` and that is absolutely a way that the same thing can be accomplished. Having both the type and default value in the same place would be easier to maintain and less error prone.

> Documenting / specifying the template's context variables similar to function arguments is an explicit goal of this proposal.
> The current proposal is more generic: you can use var in macros and anywhere inside a template.
> The var tag on the last line serves as documentation and safety net. It guarantees that, userName is always string, even though user.fullName itself is nullable.

I think there are two different things we are talking about here in regards to a template's context variables:
1. I was talking about the context variables that are passed in when rendering or including a template, so they are like "arguments".
2. You are talking about the more general context variables which encompass both those that are initially passed (that I was talking about) and the context variables as they may change during runtime via `set`, etc.

I think it is important to distinguish between the two. Using your example to expand on what I mentioned in my last comment, I think it would be more clear to augment the existing `set` tag since you are creating a new context variable:
```twig
{% set string userName = (
    user.isLoggedIn()
        ?
    user.fullName
        :
    'Anonymous'
) %}
```

If that variable could be passed to the template instead, this could be done:
```twig
{% var string userName = 'Anonymous' %}
```

**Comment by drjayvee** [comment]
> @drjayvee you are explaining how the php implementation would work. But how would the other implementations work? More generally, from a specification point of view, how would `\Foo\Bar` be resolved?

At compile time, Twig checks whether the type is valid. For classes, that would simply be done using `class_exists()`. For scalars, that's even simpler. Then again, we might let PHPStan do these checks.

For runtime (i.e., in the compiled template's PHP code), a check is added to assert the type.

For example:
```twig
{% var int number %}
{% var \User user %}
{{ user.name }} has {{ number }}
```

Would compile to something like the following, if (and only if) `strict_variables` is enabled:
```php
if (!is_int($context['int'])) {
  trigger_error('Invalid type for variable "int"'); // or throw an Exception
}
if (! $context['user'] instanceof \User) {
  trigger_error('Invalid type for variable "user"'); // idem dito
}
```

Do you see any problem here so far? Or significant limitations which render this whole feature moot?



> also it is fundamental to establish how type checking would work? Let's say the a variable x of type \Foo\Bar is used this way:
`{{ x.y.z }}`
What would be the algorithm that compilers are supposed to implement to check that x.y.z is a valid member access?

That's an interesting question. I'd hope to leverage PHPStan to do those chained checks. At the very least, though, we could check whether `y` is a valid property method of class `x` (_if_ `x`'s type is declared to be a class at all). That should be simple enough, and will already provide significant value.

However, even if properties/methods _aren't checked at all_, there's still value.

Consider this example:
```twig
{% var \Foo\Bar foo %}
{% set baz = foo.bar.getBaz() %}
{% var string baz %}
```

At compile time, we can ensure that the type is valid. The `var` tag also serves as (standardized) type documentation, as discussed in the OP.

At runtime, we can trigger warnings/errors or throw Exceptions if `foo` is not an instance of `\Foo\Bar`, and if `baz` is not a string. That alone will add a significant safety net to developers.

The goal of `var` is not to match PHPStan's level 9. Just adding what I'm proposing will add significant value. `var` is also entirely optional, so it doesn't _have to_ support complicated use cases - certainly not in the first iteration.

**Comment by drjayvee** [comment]
@willrowe 

> I think there are two different things we are talking about here in regards to a template's context variables:

Oh, do I understand correctly that you'd want the default value in `var` to apply _exclusively_ to injected variables, not ones created using `set` and macro arguments?

In that case, I think this would get too complicated, both in the implementation, as well as for users to understand.

> I think it would be more clear to augment the existing set tag since you are creating a new context variable

I guess I wouldn't be opposed to adding optional types to `set`.

However, `set` can define (the documentation reads "assigned") multiple values, and be used "to 'capture' chunks of text". This might make it more complicated to implement and support.

```twig
{% set string foo, int bar = 'foo', 1337 %} {# should work fine #}

{% set string baz %} {# pointless, as captured chunks will always be guaranteed to be string (right?) #}
    Baz
{% endset %}
```

Moreover, `set` declarations have limited scope in loops (see documentation), so the implementation would need to take that into account.

I guess _that_ could all work. But you'd need typing support in `set` _and_ default values in `var`, right?

It's an interesting idea for sure.

My gut still says it's better separate those concerns. But that's just one opinion.

**Comment by ericmorand** [comment]
@drjayvee, what about other compiler implementations like twig.js or Twing? What would be the logic that they need to implement?

**Comment by willrowe** [comment]
@ericmorand what do you currently do for [`constant`](https://twig.symfony.com/doc/3.x/functions/constant.html)?

**Comment by ericmorand** [comment]
@willrowe since there was no specification, we made an arbitrary call:

https://twing.nightlycommit.com/specifics.html#constant

But constant is an easy case to handle since the value is picked from the data passed to the context.

Types are another story and need a proper specification to be supported by non-PHP implementation.

**Comment by willrowe** [comment]
@ericmorand do you have a specific question about how it should function when a template is rendered?

You keep asking about the specification and how it should be implemented outside of PHP, but it doesn't seem like that is something to be decided here in this issue or even in this project. You may need to decide for yourself how you are going to implement it in your own project so it matches TwigPHP in a way that is consistent with how you have implemented all the other features.

**Comment by drjayvee** [comment]
@ericmorand 

> @drjayvee, what about other compiler implementations like twig.js or Twing? What would be the logic that they need to implement?

I honestly wasn't aware of these! It's interesting that there are already two.

Not to be disrespectful, but I wouldn't want to limit the capability or complicate the spec to cater to implementations other than TwigPHP, especially ones in other languages. (Given the [difficulties described here](https://twing.nightlycommit.com/specifics.html#twig-specification-implementation), this might already be the stance of the Twig maintainers.)

To answer your question directly: the easiest path is to simply ignore `var`. Alternatively, Twing (and any other implementation) could chose their own specification and implementation of the `var` tag.

[Twing's documentation](https://twing.nightlycommit.com/specifics.html#interpretation-of-the-specification) already reads:

> Twig specification being intimately bound to the TwigPHP implementation, some parts of it don’t make sense outside of the context of a PHP runtime. Below is the list of those parts that Twing interprets differently from its TwigPHP counterpart.

`var` would be very similar to `constant`: either chose a different specification and implementation, or just don't support it.

**Comment by willrowe** [comment]
@drjayvee 

> Oh, do I understand correctly that you'd want the default value in var to apply exclusively to injected variables, not ones created using set and macro arguments?

Yes, that is correct.

> In that case, I think this would get too complicated, both in the implementation, as well as for users to understand.

I do not think the implementation would be complicated, but I do agree that it would need to be done very carefully so that it is not ambiguous to the user what is happening. Naming a tag that does this `var` would probably be confusing.

> I guess I wouldn't be opposed to adding optional types to set.
> However, set can define (the documentation reads "assigned") multiple values, and be used "to 'capture' chunks of text". This might make it more complicated to implement and support.
> Moreover, set declarations have limited scope in loops (see documentation), so the implementation would need to take that into account.

I think all of this could be addressed without it becoming insurmountably complex. Using a type on a `set` block would just be disallowed. I think augmenting `set` would be very clean from a template perspective and would be familiar and easy to understand for template authors. It might get confusing to have two separate tags (`set` and `var`) that seem similar, but don't do the same thing. If you always need to set the value of the variable anyways, combining it into one tag makes a lot of sense to me.

> I guess that could all work. But you'd need typing support in set and default values in var, right?
> It's an interesting idea for sure.
> My gut still says it's better separate those concerns. But that's just one opinion.

I absolutely agree that these should be separate concerns. What I suggested as far as default values would need to be thought through very carefully and implemented as a completely separate tag that is well named so there is no confusion. I think what you have suggested here would be valuable to add. As I said above, in order to limit any confusion and make it as simple as possible, I think it should be added to the `set` tag instead of a new, separate tag.

**Comment by fabpot** [comment]
I've been thinking about this proposal for a bit now.
I think I'm ready to share my thoughts now.

My take might sound a bit harsh, but this is mainly because I care a lot about Twig philosophy (more on that in a minute).

"In short, since the PHP language itself as well as its community are moving to stricter typing, it makes sense for Twig to support this as well by adding a var tag."

I strongly disagree with this first sentence. Twig is a templating language. Twig should not become PHP. And I certainly don't want Twig to become stricter (the reasons are explained below). But, I want to consider adding some support to ease the work of static analyzers, which is how this discussion started.

Outside of static analyzers, we might optionally use that information to optimize performance (think `getAttribute()` compile time resolution) via a NodeVisitor that would not be part of Twig core (can be part of `extra` though).

And of course, the usage of the new `var` tag must be optional.

So, let me share with you a bit of **Twig's philosophy**:
The person writing Twig templates is not necessarily a developer (think templates updated by end users in a CMS backend).
They might have a limited understanding of PHP/JavaScript/... and that's fine.
So, they should not need to understand the difference between a method call, an object property, or an item of an array.
**That's the power of the `.` Twig operator.**
The `.` operator also allows to use a pure array in tests/staging/... and objects/arrays/... in production.
I should also mention that the data passed to templates should be "prepared". Most data massaging should happen before they are sent to the template (that's the reason why, for instance, mutating an array is complicated in Twig (via the `merge` filter).

With that in mind, I don't think we will support types in a template outside the `var` tag (so no support in `set` for instance).
I also think that the information from the `var` tag won't be used at compile time or runtime in core.

Starting small would be my advice here. I would focus on the `var` tag in the context of providing extra information for static analyzers, everything else can come later on.

**Comment by willrowe** [comment]
@fabpot so would this be more like front matter metadata to start with?

**Comment by smnandre** [comment]
I guess "PhpDoc" may be a good analogy or... even better for recent PHP versions: PHP attributes.

It allows to "mark" templates but would have no runtime effect.

And this would open the door for tools (static analysis beeing one) to plug-in and leverage this data.

**Comment by fabpot** [comment]
> @fabpot so would this be more like front matter metadata to start with?

No, having a proper `var` tag is fine for me.

**Comment by fabpot** [comment]
> I guess "PhpDoc" may be a good analogy or... even better for recent PHP versions: PHP attributes.
> 
> It allows to "mark" templates but would have no runtime effect.
> 
> And this would open the door for tools (static analysis beeing one) to plug-in and leverage this data.

That's indeed a good analogy.

The first idea (and the one I think implemented by VS Code) is to use comments. And this proposal is about having the proper infrastructure at Twig level so that everybody can share the same (like PHP attributes).

**Comment by drjayvee** [comment]
@fabpot thanks for your thoughtful response. I agree completely that typing should always be optional, and that Twig must not become PHP (let alone at max PHPStan).

I've had a chat with @ruudk today, and he showed the progress he made to leverage PHPStan to do inspection on compiled templates. I think there's great promise in that approach.

What is missing in Twig to enable both this or my humbler approach is a way to specify types. The PHPDoc route of using comments was rejected (see #4009) to keep them out of the AST. Therefore, a proper tag is the only way to go.

As I've pointed out before, the community is already using `@var` _comments_ to declare types. Therefore, a proper `var` tag is an easy migration.

By the way, I'm working on adding the `var` tag to the TwigStan extension. However, I do think it should be an official part of Twig so that the community adopts it.


On a different note, using FQNs in `var` breaks the Lexer due to the leading `\`. This easily (though naively) solved by adding that character to `Lexer::PUNCTUATION`, but adding more complicated types (e.g., `array{foobars: array<\Foo\Bar>}`) will probably cause more breakage.

We could consider extending the Lexer, I'd very much prefer to keep the required changes to Twig core to a minimum.

Therefore, I propose types must always be quoted. I'll amend the OP.

**Comment by fabpot** [comment]
@drjayvee It looks like we are very much aligned :) Let's add a `@var` tag to make it the official "solution" (like PHP added attributes).

**Comment by drjayvee** [comment]
Awesome! I'll start working on an implementation and documentation.

I do think Twig should at least validate types to prevent mistakes. We can leave other inspections to extensions. Would you agree, @fabpot?

@ruudk's prototype had a way to specify multiple types in one go, which we could consider here as well:
```twig
{% var {
  foo: 'string',
  bar: '\Foo\Bar'
} %}
```

This would make the spec and implementation more complicated, plus the order of variable name and type is flipped in this notation. Therefore, I'd prefer to keep it simple and only support what's currently written in the OP.

**Comment by fabpot** [comment]
> @ruudk's prototype had a way to specify multiple types in one go, which we could consider here as well:
> 
> ```
> {% var {
>   foo: 'string',
>   bar: '\Foo\Bar'
> } %}
> ```
> 
> This would make the spec and implementation more complicated, plus the order of variable name and type is flipped in this notation. Therefore, I'd prefer to keep it simple and only support what's currently written in the OP.

The implementation is going to be trivial as this is just an expression. So, a call to`ParseExpression::parseMappingExpression()` to enforce a mapping will do all the work.

**Comment by drjayvee** [comment]
So let's (try to) support formats:
```twig
{% var 'string' foo %}
{# and #}
{% var {
  bar: 'string',
} %}
```

**Comment by fabpot** [comment]
> So let's (try to) support formats:
> 
> ```
> {% var 'string' foo %}
> {# and #}
> {% var {
>   bar: 'string',
> } %}
> ```

Why not support just 1 format?

**Comment by drjayvee** [comment]
Well, I don't love this format for single variables.

But then, most of the times (start of template and start of macro), multiple variables will need to be typed.

So sure, let's do the multi-type mapping expression exclusively then. I'll adapt the OP now and start working on an implementation soon.

**Comment by stof** [comment]
but then, is it an issue to have multiple `{% var %}` tags at the beginning of the template if we consider the single-variable format more readable ?

**Comment by stof** [comment]
Note that `parseMappingExpression` would allow complex expressions in values of the mapping, which we probably don't want to support in the `{% var %}` tag, so I don't think we can reuse it directly.

**Comment by fabpot** [comment]
> Note that `parseMappingExpression` would allow complex expressions in values of the mapping, which we probably don't want to support in the `{% var %}` tag, so I don't think we can reuse it directly.

Correct. Let's copy/paste and simplify.

**Comment by willrowe** [comment]
I know everyone seems pretty locked into the name being `var` since it is taken directly from PHPDoc, but from the perspective of whoever is writing the templates (who may not be a PHP developer), could it be seen as ambiguous?
For example, if I'm a JavaScript developer and I look at a Twig template and see the `set` and `var` tags, I'm going to think `var` is creating a variable and would not necessarily think `set` would do that.

I just want to make sure this is taken into account, I don't necessarily feel strongly that this would be a big issue, but might as well make sure the name is clear from the outset.

**Comment by fabpot** [comment]
> I know everyone seems pretty locked into the name being `var` since it is taken directly from PHPDoc, but from the perspective of whoever is writing the templates (who may not be a PHP developer), could it be seen as ambiguous? For example, if I'm a JavaScript developer and I look at a Twig template and see the `set` and `var` tags, I'm going to think `var` is creating a variable and would not necessarily think `set` would do that.
> 
> I just want to make sure this is taken into account, I don't necessarily feel strongly that this would be a big issue, but might as well make sure the name is clear from the outset.

That's a very good point. Let's think about possible alternatives (that does not block the implementation anyway).

**Comment by ruudk** [comment]
I'm also working on a static analyzer for Twig. Next to supporting @var I came up with `{% type_hint name 'string' %}` as a way to hint type information to the analyzer. It's completely ignored on compilation/runtime.

**Comment by stof** [comment]
I would rather use `{% type %}` than `{% type_hint %}`

**Comment by willrowe** [comment]
I think `{% type %}` is very clear, no ambiguity there. If it will be the multi-variable syntax then plural `types` may be even better.
```twig
{% types {
  bar: 'string',
  foo: 'bool',
} %}
```

```twig
{% types {
  bar: 'string',
} %}
```

**Comment by ruudk** [comment]
> I would rather use `{% type %}` than `{% type_hint %}`

Yeah that would also be fine

**Comment by drjayvee** [comment]
I chose `var` to match PHPDoc's `@var`, but I do prefer `types` (plural) thanks to improved clarity.

Given Fabian's vision that Twig template developers don't necessarily know PHP, I don't think consistency with PHPDoc is more important than ease of use.

**Comment by fabpot** [comment]
I like `types` as well.
To be honest, I don't think non-developers will use the new tag anyway ;)

**Comment by drjayvee** [comment]
> but then, is it an issue to have multiple {% var %} tags at the beginning of the template if we consider the single-variable format more readable ?

@stof is your question still relevant given that we've decided for multi-variable `types`?

To clarify, I think this tag should be allowed anywhere inside a template. Two `types` right next to each other are valid, even though it's always better to combine them.

```twig
{% types {foo: string} %}
{% types {bar: string} %}

{% set baz = randomMixedValue() %} {# PHP return type is mixed #}
{% types {baz: string} %} {# but we're declaring that it will be string here #}
```

**Comment by smnandre** [comment]
I think I slightly prefer the _one-liner_ over the _json format_ in terms of readability ..

### One var

```twig
{% type App\Entity\Post[] posts %}
```

```twig
{% types {posts: App\Entity\Post[]} %}
```

### Two vars


```twig
{% type posts  App\Entity\Post[] %}
{% type bar    App\Foo\Bar %}
```

```twig
{% types {
    posts: App\Entity\Post[],
    bar: App\Foo\Bar,
} %}
```

**Comment by willrowe** [comment]
The issues I run into the most with templates are maintainability and consistency. To help with these issues, I would only ever use one single `types` tag at the top of the file and it would include every variable in it. That way there would be no searching around the file for other stray tags or accidentally including a variable more than once.

Even if there is only one variable in the template, the format would still be the same and I could add another line to it if another variable is needed. If the tag looks the same in every template file, then it is easier to cognitively parse. This is just my personal preference because I prefer consistency over having more than one way to do something based solely on how it looks visually.

**Comment by ruudk** [comment]
> I think I slightly prefer the _one-liner_ over the _json format_ in terms of readability ..

@smnandre Just an idea: We could do the following:

```twig
{% type name = 'null|string' %}
```

And if you want to define multiple, you can do it in a one liner:
```twig
{% type name = 'string', users = 'array<int, App\User>' %}
```

Or define 2 separate lines:
```twig
{% type name = 'string' %}
{% type users = 'array<int, App\User>' %}
```

We could make `types` an alias of `type`, so that you can use `types` for when working with multiple types.

> [!NOTE]
> I think it's important that the type representation is always a quoted string. This will allow for more complex PHPDoc style types like `null|string` and `null|array<int, App\User>`. Without this being quoted, we need to change the Lexer to allow for all these special chars and I don't think that's ideal.

> Even if there is only one variable in the template, the format would still be the same and I could add another line to it if another variable is needed. If the tag looks the same in every template file, then it is easier to cognitively parse. This is just my personal preference because I prefer consistency over having more than one way to do something based solely on how it looks visually.

@willrowe While it's not a blocker for me, I do think that it would be great if we could standardize this and say that there can only be 1 type(s) tag in a given scope (top of the template and per macro body). This would make it easier for static analyzers to see what kind of requirements a given template has.

**Comment by smnandre** [comment]
Concerning the quote/not..  and order type-name / name-type... i have no opinion :) 

But you're 100% right about the Lexer!
Using quotes will be much easier... so much easier that, in fact, there nothing to change :) 

```php
    public function testFooBar()
    {
        $template = "{% type '@Foo\\Bar☘️' %}";

        $lexer = new Lexer(new Environment(new ArrayLoader()));
        $stream = $lexer->tokenize(new Source($template, 'index'));

        $stream->expect(Token::BLOCK_START_TYPE);
        $this->assertSame('type', $stream->expect(Token::NAME_TYPE)->getValue());
        $this->assertSame("@Foo\Bar☘️", $stream->expect(Token::STRING_TYPE)->getValue());
    }
 ```
 
**But** I have some reservations about using ` = ` in this context, as it ~~could potentially~~ **will 100%** cause confusion with variable assignment... it may not be the best choice for developer/user experience.

Especially since we're gradually transitioning to using `  :  `  for named arguments.

**Comment by fabpot** [comment]
Maybe we can use `as` instead of `=`?

**Comment by smnandre** [comment]
> (top of the template and per macro body)

What bout blocks ? (`embed`, `use` ..)  And variables set in the template ? 

As i imagine it (as a "docblok/attribute").. i see no reason to forbid multiple tags in a template .. 🤷‍♂️   

and of course..  if you want to follow a "only-one" rule.. no one would force you to do differently ;)

**Comment by smnandre** [comment]
> `as`

 `As` is frequently used for aliasing / imports  .. this may not be ideal for front-end developers for instance :|

**Comment by ruudk** [comment]
> Maybe we can use `as` instead of `=`?

To visualize it:

```twig
{% type name as 'null|string' %}
```

```twig
{% types name as 'string', users as 'array<int, App\User>' %}
```

**Comment by ruudk** [comment]
Another idea, keeping it simple:

```twig
{% type name 'null|string' %}
```

```twig
{% types name 'string', users 'array<int, App\User>' %}
```

**Comment by smnandre** [comment]
> To be honest, I don't think non-developers will use the new tag anyway ;)

Hmm... I'm sorry to add another suggestion, just thinking out loud. 


If this is meant _for_ developers _by_ developers...
.... I wonder if using `assert` might be more conventional (and future-proof)?

```twig
{% assert search_query 'string' %}
{% assert page_num 'positive-int' %}


```

**Comment by smnandre** [comment]
Because more than just "type / statical analysis" ... this could be used as "contract / validation" in template

```twig
{% assert search_query|length > 0 %}
{% assert size in ['red', 'blue'] %}
```

This could leverage all the existing expressions / tests / etc  ... 

I'm sorry this was not the original topic 😶 

But i'll note the idea and see what can be done with this, because for components --in general, not only UX ones-- this could be a very helpfull feature

**Comment by stof** [comment]
I would say that `assert` and `types` are 2 different features that could both be implemented. Using assertions to derive types would require having type-checking expressions in Twig, which would be much harder than having this `types` tag documenting the expected type.

**Comment by drjayvee** [comment]
I agree with @ruudk that

1. Users can always use multiple one-line tags if they prefer. This has the benefit of affording a comment next to the variable type.
2. Adding an alias `type` for `types` could be useful. Then again, I prefer simplicity, and don't think the plural as the only option is that problematic.

> While it's not a blocker for me, I do think that it would be great if we could standardize this and say that there can only be 1 type(s) tag in a given scope (top of the template and per macro body)

Here I do disagree. I think it can be very useful to document (_and assert_, more on that later) types anywhere. See [my previous example](https://github.com/twigphp/Twig/issues/4165#issuecomment-2306591773). I'd be happy to create a more elaborate and realistic example, but I think you get the point.

> Concerning the quote/not.. and order type-name / name-type... i have no opinion :)

I do! I've tried adding support for just FQN to Twig, but the Lexer doesn't support it. We'd have to make significant changes.

I've [already implemented](https://github.com/alisqi/Twig/blob/add-types-tag/src/Node/TypesNode.php) (WIP, obviously) a `TypesNode` which supports and asserts simple mappings (`{name: 'string'}` ), and that was very easy. No changes to Twig are needed, only the addition of the `TypesNode`.

However, I do have one other issue: validation of types. I would love for Twig core (not extensions like TwigStan) to validate types. We can certainly implement validation for the limited type syntax in the OP. However, @ruudk has shown me how powerful full PHPStan support can be. Unfortunately, asserting those types would require taking on PHPStan itself as a dependency, which surely is out of scope. Re-implementing the full syntax (perhaps by copying PHPStan code) is also not a great idea.

We could say that the type string can be anything (as long as it's a _constant string_). But that would severely limit the utility of `types` because there's no guarantee (out of the box) that the type itself is valid (e.g., `boo` instead of `bool`), let alone the variable's run-time type.

Another problem is that I'd love (see OP) optional _run-time assertions_ (see below) added to compiled templates based on `types`. I don't know PHPStan's internals, but I doubt whether it could do those, since that project is all about static analysis.

What do you all think? Perhaps the official `types` should support any type supported by PHP itself, and leave extensions (such as PHPStan's full array of types (pun fully intended)) to extensions (idem dito re: pun).

> I would say that `assert` and `types` are 2 different features that could both be implemented. Using assertions to derive types would require having type-checking expressions in Twig, which would be much harder than having this types tag documenting the expected type.

I fully agree here and have added a comment to @smnandre's issue. One important thing I'll copy here is that my intention/goal is to have assertions on variable types handled by TwigStan or maybe Twig itself.

**Comment by stof** [comment]
Runtime type validation would add overhead, especially for complex types. And it would be quite hard if we want to support generics as well (as those don't exist on the PHP side).
That's why leaving the validation of types to static analyzers is better to me than trying to turn them into runtime assertions, which would limit the usable types to the ones supported by the Twig type system instead of supporting anything the static analyzer supports.

Btw, Fabien already agreed on working on this simpler scope, not on adding a runtime type system in Twig.

**Comment by drjayvee** [comment]
> Runtime type validation would add overhead, especially for complex types.

Agreed, which is why I'm proposing to only enable those if `strict_types` is enabled. That's probably only used in dev environments, where the overhead is less important than code quality.

> That's why leaving the validation of types to static analyzers is better to me than trying to turn them into runtime assertions, which would limit the usable types to the ones supported by the Twig type system instead of supporting anything the static analyzer supports.

Full agree here. I just did implement validation for simple types. But that implementation would then throw on more complicated ones, severely limiting what extensions can do.

So, while most analyzer will most likely support [PHPDoc/PHPStan](https://phpstan.org/writing-php-code/phpdoc-types) (and use PHPStan under the hood), Twig itself will stay mum on the syntax and spec of types.

> Btw, Fabien already agreed on working on this simpler scope, not on adding a runtime type system in Twig.

Fair enough.

~~I'll adjust~~ I've adjusted the OP accordingly.

**Comment by stof** [comment]
> only enable those if `strict_types` is enabled. That's probably only used in dev environments, where the overhead is less important than code quality.

I can tell you that this assumption is wrong.
I found out that enabling `strict_types` in prod was actually useful:

- it has no overhead compared to disabling `strict_types`
- the fallback to `null` does not help:
    - many filters or functions will fail when receiving `null` as input, but the stack trace will not allow you to find the root cause of the issue as it won't show where the `null` comes from
    - some functions or filters will work but produce unexpected results when getting `null`. For instance, the `date` filter (and its siblings from `twig/intl-extra`) convert `null` to `new \DateTimeImmutable('now')`, which might not be what you expect.

**Comment by drjayvee** [comment]
> many filters or functions will fail when receiving null as input, but the stack trace will not allow you to find the root cause of the issue as it won't show where the null comes from

That's the very problem that lead me to investigate static analysis for twig. I had renamed a variable (in a macro) and forgot to change one of the uses. It broke the whole page in production. This is no longer possible with [our extension](https://github.com/alisqi/TwigStan).

More generally though, we could reconsider the utility for `strict_variables` (not `strict_types`)? But that's an issue for another time.

**Comment by drjayvee** [comment]
In PR #4235, the documentation for `types` includes this note:

> The syntax for and contents of type strings are intentionally left out of scope.

@fabpot added the following in his review (merci beaucoup pour ca!)

> Thinking out loud here: maybe we can document a default subset that should always be supported by extensions, the basic types used by Twig: boolean, number (maybe integer, float), mapping, sequence, string/stringable, object, iterable?

I think this is a good idea, but I have some reservations. Let me also think out loud here.

It's weird that Twig core specifies should/what must be supported, while explicitly leaving validation and support for more types to said extensions.

Then again, it would be great if the basics are specified to give the tag more weight and possibly direction.

I'd propose `bool`, `string`, `float`, `int`, `array`, `object` and `mixed`. (`null` is pointless, right?)

While I like `number`, I prefer `int` and `float` because those are the actual types in PHP, plus the distinction conveys relevant information.

Next, while I like `mapping` and `sequence`, I'm afraid those will need their own detailed specifications and implementations in extensions. Most likely is that implementations will use PHPStan's types, as those are most widely used and supported by other tools, including IDEs.

What's more, specifying `mapping` but not its keys and values' types, and specifying `sequence` without the items' type(s) adds very little over `array`.

Lastly, `mixed` is necessary unless we support union types (e.g., `int|string`), which goes way to far for core. Without `mixed` you can't document that a variable is required (unless it's of one of the other types.)

I would also really like support for nullable by using the `?` prefix to the type.

I predict we'll see lots of different opinions. Which isn't to say we shouldn't have it and stick to just agnosticism. Should we consider this a separate issue? I guess we should either try to merge the tag _with_ a basic list of types, or decide to remain completely agnostic.

**Comment by stof** [comment]
Documenting a default subset would mean that extensions wanting to validate type would be interoperable on the meaning of this default subset, making things more useful. See for instance `bool` vs `boolean` which would hurt interoperability if 2 extensions decide to use different names for the boolean type.
However, we could also let extensions collaborate between themselves as a start to see how things go (for instance, in the PHP land, psalm and phpstan tend to harmonize the types they support for the same reason)

**Comment by JoshuaBehrens** [comment]
Regarding get_attribute optimization: I would hate to work with mixed just because someone has an "`object|array` kind of situation" because their code wants to support 1) a "pagination"-object they get or 2) someone fakes the object by passing an array in the `include … with { pagination:  }`. Then everyone will just use mixed. And therefore I cannot predict to go for array access or method call check :/ 

Do I get it right, that any variable is expected to be nullable? So there is no need to write `string|null`?

The pagination DTO is just something to make it less abstract. Insert any case you have in your mind, where a template requires certain data to be used.

**Comment by willrowe** [comment]
> Do I get it right, that any variable is expected to be nullable? So there is no need to write string|null?

I would like the ability to explicitly mark variables as accepting a null value or completely nullable. To give some examples, I currently document variables that are optional and are not required to be passed to the template context like so:
```twig
{% types {
    optional_variable: '?boolean',
} %}
```

Usually, I will then set that variable to a default value if it has not been passed. The `?` denotes that it is not required. This is distinct from allowing a `null` value to be passed to a variable though:
```twig
{% types {
    optional_variable: 'string|null',
} %}
```

In this case, if some text should be displayed in an `h1` tag, then it should be passed as a string. If the `h1` tag should be omitted altogether then `null` should be passed explicitly.

**Comment by fabpot** [comment]
Can we say that a variable is optional if there is a `is defined` test or a `default` filter? The more tools can deduce the better IMHO.

**Comment by willrowe** [comment]
@fabpot that's a great question. I'm only thinking in terms of what is passed in the template context.

**Comment by ruudk** [comment]
If you think of the context as an array, in PHP you would type optional array keys as follow:

`array{key: 'string', optionalKey?: 'string'}`

See https://phpstan.org/writing-php-code/phpdoc-types#array-shapes

It would be great if we could write:
```twig
{% types {
  key: 'string',
  optionalKey?: 'string'
} %}
```

Also, if we want to describe what are valid values for types, I would say: any type you would write in PHPDoc. 

See https://phpstan.org/writing-php-code/phpdoc-types

**Let's keep it simple** and see what analyzers do with it. We can standardize things depending on use cases in the wild. After things are proven to be useful.

**Comment by willrowe** [comment]
@ruudk that's an interesting approach. I'm not picky about the syntax, I would just like a way to denote it that is valid and doesn't conflict with anything else.

**Comment by smnandre** [comment]
I'm thinking too the base types of twig should be "documented" to encourage compatibility. 

For the rest, i kinda feel this could be solved by _usage_ pretty quickly, as users will do what they _need_, tools too, and thus can be standardized in a future step.

**Comment by drjayvee** [comment]
**Optional variables**
> Do I get it right, that any variable is expected to be nullable? So there is no need to write string|null?

My idea with the `?` prefix is to only use it as shorthand for `|null`, just like in PHP. By default, variables are not considered nullable. (Again, that's my proposal.)

I would really encourage everyone (who's interested in `types` and code quality) to ensure variables are _always defined_.  But that might be too opinionated an approach for Twig core. Alternatively, users can just leave optional variables out of `types`, of course, but that creates a gap in what can and cannot be declared. :/

> Can we say that a variable is optional if there is a is defined test or a default filter? The more tools can deduce the better IMHO.

I prefer ruud's proposal (i.e., `{name?: 'type'}`) to declare which variables are optionally defined. Otherwise, template developers would have to go looking for `name is defined` in the entire template, so there's way less clarity.

(Edit: I just tried adding the implementation for `name?` to my PR branch, and it turns out to be trivial. 🚀)

(Extensions can trigger a warning if an optional variable is used without `is defined` - at least in theory. They can also trigger a warning for optional variables and advise to ensure variables are defined.)

Do we all agree that we do want support for optional variables in `types`? (I do!)

Here's an example:
```twig
{% types {
  requiredInt: 'int',
  optionalInt?: 'int',
  requiredNullableInt: '?int',
  optionalNullableInt?: '?int',
} %}
```

**Types**
It's good to see people agree that defining a base set of types is useful.

So do you prefer matching PHP's concrete types (e.g., 'string, array`), or Fabian's more high-level ones (e.g., 'stringable, iterable')?

Again, extensions can add types and validate types as they see fit. I'm okay with extension developers figuring out interoperability amongst each other. Most likely, extensions will settle on PHPStan/Psalm.

**Comment by stof** [comment]
I like the `?` suffix on the variable name to indicate optional variables. That's a nice syntax, and it is consistent with the syntax for array shapes used in phpstan and psalm.

For the basic types, I would say that the concrete PHP types need to be part of the list, but this list might include other types. For instance, `iterable` makes perfect sense. But anyway, this can happen later compared to adding the `{% types %}` as it is purely documentation aimed at type-checking extensions. The priority is introducing the tag so that those extensions can start consuming it.
And I indeed expect those extensions to use types matching the phpstan/psalm ones (especially as it can allow them to actually use phpstan/psalm in their implementation). This will be easier for devs that won't need to learn 2 different ways to describe types. But let's see how those extensions evolve.

**Comment by ruudk** [comment]
> For instance, `iterable` makes perfect sense.

And that's perfectly valid according to [PHPStan's documentation](https://phpstan.org/writing-php-code/phpdoc-types#basic-types)

**Comment by ruudk** [comment]
> My idea with the `?` prefix is to only use it as shorthand for `|null`, just like in PHP. By default, variables are not considered nullable. (Again, that's my proposal.)
> {% types {
  requiredInt: 'int',
  optionalInt?: 'int',
  requiredNullableInt: '?int',
  optionalNullableInt?: '?int',
} %}

That's not a valid type in PHPDoc. In PHPDoc you write `string|null`.

So it would become:
```twig
{% types {
  requiredInt: 'int',
  optionalInt?: 'int',
  requiredNullableInt: 'int|null',
  optionalNullableInt?: 'int|null',
} %}
```

---

I think to support the `?` at the end of a key we need to modify `Twig\ExpressionParser::parseMappingExpression` to support it. Maybe we can create a new one that parses static mappings. So that means that keys can end with `?`. And that values should always be a string (not an expression).

Twig also supports shorthands like `{ name, other: 'value' }` where `name` is an existing variable that is then written as `name: name`. Just like modern JavaScript/TypeScript. In the case of this `types` mapping, it should not be possible.

**Comment by fabpot** [comment]
To give us more time to experiment, maybe we could mark the new tag as being "experimental" like what we do on Symfony. That will let us change the details after we gain enough experience using it.

**Comment by drjayvee** [comment]
> [?int] is not a valid type in PHPDoc.

Right, but it _is_ valid PHP. 🤕 I actually kind of prefer `|null` anyway because it's more explicit.

> I think to support the ? at the end of a key we need to modify Twig\ExpressionParser::parseMappingExpression to support it.

Nope. [My current implementation](https://github.com/alisqi/Twig/blob/ceb47491938cc81c0a09e491cafad14972284203/src/TokenParser/TypesTokenParser.php#L51) doesn't use that method because it supports way more than `types` should, such as the `{key}` shorthand for `{key: key}`, expressions for values, and more.

Adding support for `name?` is trivial:
```
diff --git a/src/TokenParser/TypesTokenParser.php b/src/TokenParser/TypesTokenParser.php
index 2bbe183f..b252b848 100644
--- a/src/TokenParser/TypesTokenParser.php
+++ b/src/TokenParser/TypesTokenParser.php
@@ -68,7 +68,10 @@ final class TypesTokenParser extends AbstractTokenParser

             $nameToken = $stream->expect(Token::NAME_TYPE);
             $nameExpression = new NameExpression($nameToken->getValue(), $nameToken->getLine());
+            
+            $isOptional = $stream->nextIf(Token::PUNCTUATION_TYPE, '?') !== null;
+            $nameExpression->setAttribute('is_optional', $isOptional);
             $stream->expect(Token::PUNCTUATION_TYPE, ':', 'A mapping key must be followed by a colon (:)');

             $valueToken = $stream->expect(Token::STRING_TYPE);
```

> To give us more time to experiment, maybe we could mark the new tag as being "experimental" like what we do on Symfony. That will let us change the details after we gain enough experience using it.

Sure, I'm totally fine with that.

**Comment by ruudk** [comment]
> [My current implementation](https://github.com/alisqi/Twig/blob/ceb47491938cc81c0a09e491cafad14972284203/src/TokenParser/TypesTokenParser.php#L51) doesn't use that method because it supports way more than `types` should, such as the `{key}` shorthand for `{key: key}`, expressions for values, and more.
> 
> Adding support for `name?` is trivial:

That's great 👏

**Comment by drjayvee** [comment]
And it's merged! Thank you all for contributing.

Do we still want to continue discussing the base types that should be supported? Do we continue in a new issue and close this one? Or do we wait and see what the community does with this tag?

I still think it would be valuable to have the basics covered and part of the official spec, but I'd prefer continuing this discussion in a new issue.

Lastly, I wanted to ask you, @Haehnchen, whether you're going to add support for `types` in the Symfony Support plugin. I've also asked the JetBrains PHPStorm team to consider [adding 1st-party support](https://youtrack.jetbrains.com/issue/WI-76637/Twig-types-tag).

**Comment by fabpot** [comment]
Thank you all for your great discussion here. Let’s open another issue for the basic types.

**Comment by drjayvee** [comment]
I'll close this issue. See you in #4256!
